Corona-Warn-App/src/main/java/de/rki/coronawarnapp/task/TaskController.kt

```diff
@@ -0,0 +1,219 @@
+            .filter { it.deferred.isCompleted && it.executionState != TaskState.ExecutionState.FINISHED }
+            .forEach { data ->
+                val error = data.deferred.getCompletionExceptionOrNull()
+                val result = if (error == null) {
+                    data.deferred.getCompleted()
+                } else {
+                    Timber.tag(TAG).e(error, "Task failed: %s", data)
+                    null
+                }
+
+                this[data.id] = data.copy(
+                    completedAt = timeStamper.nowUTC,
+                    result = result,
+                    error = error
+                )
+                Timber.tag(TAG).i("Task is now FINISHED: %s", data)
+            }
+
+        // Start new tasks
+        values.toList()
+            .filter { it.executionState == TaskState.ExecutionState.PENDING }
+            .forEach { data ->
+                Timber.tag(TAG).d("Checking pending task: %s", data)
+
+                val siblingTasks = values.filter {
+                    it.type == data.type &&
+                        it.executionState == TaskState.ExecutionState.RUNNING &&
+                        it.id != data.id
+                }
+                Timber.tag(TAG).d("Task has %d siblings", siblingTasks.size)
+                Timber.tag(TAG).v(
+                    "Sibling are:\n%s", siblingTasks.joinToString("\n")
+                )
+
+                // Handle collision behavior for tasks of same type
+                when {
+                    siblingTasks.isEmpty() -> {
+                        this[data.id] = data.toRunningState()
+                    }
+                    data.config.collisionBehavior == CollisionBehavior.SKIP_IF_ALREADY_RUNNING -> {
+                        this[data.id] = data.toSkippedState()
+                    }
+                    data.config.collisionBehavior == CollisionBehavior.ENQUEUE -> {
+                        Timber.tag(TAG).d("Postponing task %s", data)
+                    }
+                }
+            }
+
+        Timber.tag(TAG).v("Tasks after processing: %s", this.values)
+    }
+
+    private fun InternalTaskState.toRunningState(): InternalTaskState {
+        deferred.invokeOnCompletion {
+            Timber.tag(TAG).d("Task ended (type=%s, id=%s)", type, id)
+            taskScope.launch { processMap() }
+        }
+        task.progress.onEach {
+            Timber.tag(TAG).v("$task progress: $it")
+        }.launchIn(taskScope)
+
+        deferred.start()
+        return copy(startedAt = timeStamper.nowUTC).also {
+            Timber.tag(TAG).i("Starting new task: %s", it)
+        }
+    }
+
+    private fun InternalTaskState.toSkippedState(): InternalTaskState {
+        Timber.tag(TAG).d("Task was skipped (type=%s, id=%s)", type, id)
+        return copy(
+            startedAt = timeStamper.nowUTC,
+            completedAt = timeStamper.nowUTC
+        ).also {
+            Timber.tag(TAG).i("Starting new task: %s", it)
```

Comment:
Copy-paste mistake, should be "Skipping task", right?

Reply:
Yep :sweat_smile: